controllers/crds/cninode_controller.go

```diff
@@ -0,0 +1,194 @@
+					}
+					// Return err if failed to delete leaked ENIs on node so it can be retried
+					if err := cleaner.DeleteLeakedResources(); err != nil {
+						return ctrl.Result{}, err
+					}
+					// Remove finalizer and update CNINode to continue deletion
+					if ok := controllerutil.RemoveFinalizer(cniNode, config.NodeTerminationFinalizer); !ok {
+						r.Log.Info("failed to remove finalizer for CNINode, requeuing the request", "cniNode", cniNode.Name)
+						return ctrl.Result{Requeue: true}, nil
+					}
+					if err := r.Client.Update(ctx, cniNode); err != nil {
+						// Return err so update can be retried
+						return ctrl.Result{}, err
+					}
+				} else {
+					r.Log.Info("failed to get the node object in CNINode reconciliation, will retry")
+					// Requeue request so it can be retried
+					return ctrl.Result{}, err
+				}
+			} else {
+				// node exists, do not run the cleanup routine(periodic cleanup routine will anyway delete leaked ENIs), remove the finalizer
+				// and let object be deleted and re-create similar object
+
+				// Create a copy without deletion timestamp for creation
+				newCNINode := &v1alpha1.CNINode{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:            cniNode.Name,
+						Namespace:       "",
+						OwnerReferences: cniNode.OwnerReferences,
+						// TODO: should we include finalizers at object creation or let finalizer add it on Create event?
+						Finalizers: cniNode.Finalizers,
+					},
+					Spec: cniNode.Spec,
+				}
+				if ok := controllerutil.RemoveFinalizer(cniNode, config.NodeTerminationFinalizer); !ok {
+					r.Log.Info("failed to remove finalizer for CNINode, requeuing the request", "cniNode", cniNode.Name)
+					return ctrl.Result{Requeue: true}, nil
+				}
+				if err := r.Client.Update(ctx, cniNode); err != nil {
+					// Return err so update can be retried
+					return ctrl.Result{}, err
+				}
+
+				// Re-fetch CNINode to verify it is deleted before re-creating it as we use the same name
+				// TODO: do we need to wait till CNINode deletion here?
+				if err := r.Client.Get(ctx, req.NamespacedName, cniNode); err != nil && errors.IsNotFound(err) {
+					if err = retry.OnError(retry.DefaultBackoff, func(error) bool {
+						if err != nil {
+							return true
+						}
+						return false
+					}, func() error {
+						r.Log.Info("creating CNINode after it has been deleted as node still exists", "cniNode", newCNINode.Name)
+						return r.Client.Create(ctx, newCNINode)
+					}); err != nil {
+						// The CNINode is deleted at this point, we cannot re-create the same object
+						// raise node object to publish warning that CNINode is deleted
+						utils.SendNodeEventWithNodeName(r.K8sAPI, node.Name, utils.CNINodeDeleted,
+							fmt.Sprint("CNINode was deleted and could not be re-created by controller"), v1.EventTypeWarning, r.Log)
+
+					}
+					// Requeue request to ensure the state of the object
+					return ctrl.Result{Requeue: true}, nil
```

Comment:
Can you add a metrics on recreating the resource for protection? I want to know how many customers may trigger this behavior. I am not really preferring to requeue this after retries on a supposed to be reliable API call. Having it for now is ok to me. I would like to remove the requeueing after we have metrics data from customers and prove the behavior should be very rare and the retry succeed.